### ISSUE TITLE
UI button to "see all conversation" using context API

### ISSUE THREAD

**Issue by twitchy-ears** [mastodon/mastodon#18150]
### Pitch

I think this has been talked about before in threads like https://github.com/mastodon/mastodon/issues/14017 and https://github.com/mastodon/mastodon/issues/3794

Essentially when viewing a toot on a small instance you are often missing replies because your instance never saw them - this is totally reasonable and expected behaviour, once you get how AP works. 

However if you view the web version of the original-toot and search for the URLs of a few reply-toots then your server knows about them and you can then see them in your view of the conversation around the original-toot.  Again very reasonable and expected behaviour.

What I am suggesting is an UI button of some kind that trips an API call (so it can be done from 3rd party clients) for my instance to contact the originating toots instance and ask for the URLs of all replies to the original-toot.  Then essentially my instance does a search on them, adding them to its pool of known toots and finally refreshes the page I'm viewing - hence they appear and I have an up to date view of the conversation from inside my own instance and can easily boost/favourite/reply to them.

Retrieving toots from remote servers clearly works, and there is already [GET /api/v1/statuses/:id/context](https://docs.joinmastodon.org/methods/statuses/#parent-and-child-statuses) which presumably could be used to pull in the conversation so I think the only thing missing is a piece of UI glue to trigger this all to happen and/or adding an API call that binds all this behaviour together (which would make it easier for 3rd party clients) and then add a UI button that sits on top of that single "fetch context" or "see all conversation" API call.

Additionally I think at least a subset of people would find it useful/accessible to have a per-profile option to automatically make this API call (if its not been called in the past X minutes for this toot) when viewing a single toot, as that way they can know they are getting an up date view of conversation on the fedi around that toot.

### Motivation

- Removes behaviour that is confusing for new or less technical users.
- Makes for a more actual and global view of conversation flow be available on small instances.  
- Removes the need for constantly going to the web view of a remote instance, scanning the replies, then working out where to hit reply from there, getting redirected back to your original instance in a popup to type the reply, etc.etc.etc.

**Comment by trwnh** [comment]
i think it might be a decent idea to crawl the `replies` property on the activitypub object (although i'd like to see #10916 dealt with if this is done)

**Comment by valerauko** [comment]
this shouldn't be hard to implement. add a button that triggers an async process to fetch a reasonable number of the statuses in the conversation or the person's public outbox, and show a spinner on the UI while it's in progress so multiple users won't try to trigger the same refetch.

that or if you're brave just passively check if there are any unknown messages in the conversation/outbox whenever someone accesses the page (or once every day for example)

**Comment by trwnh** [comment]
big problem: you can't reliably or consistently paginate outboxes

**Comment by valerauko** [comment]
>you can't reliably or consistently paginate outboxes

is that so? i quickly checked a few major implementations and mastodon, misskey and pleroma all have an OrderedCollection for the outbox that can be paginated

**Comment by trwnh** [comment]
@valerauko you can paginate but not reliably or consistently. which page do you start at? how do you construct your request? these things are not standardized, so you'll end up special-casing for every single implementation at best.

for example:

- mastodon does `https://mastodon.example/users/username/outbox?max_id=109344512179163590&page=true` so you need to extract that the mastodon-local ID from the activitypub id of the last status you have (something like `https://mastodon.example/users/username/statuses/109344512179163590` probably). you also can't detect any gaps in the middle of what you already have, so working backwards is all you can do reliably.
- pleroma does mostly the same as mastodon, but ids are Flake IDs. you still can't detect gaps, you still have to make the same assumptions. except there's an extra nickname parameter: `https://pleroma.example/users/a/outbox?max_id=APSWvVvX80JpPYYKpc&nickname=a&page=true` which we have to special-case for pleroma (and we don't know which servers are running pleroma, and we shouldn't care from a spec perspective)
- misskey uses hash ID and paginates by since_id and until_id, for example `https://misskey.example/users/5c4d7bf7cacf1f25ec02f97c/outbox?page=true&until_id=81847802b4c5a40cfc84be17` -- so now we have to special-case misskey, when we don't know (and shouldn't care) when some server is running misskey
- any other server is not bound to do the same or even similar conventions as the three big ones described above which already vary. how does gotosocial work? pixelfed? friendica? and so on
- actually i know how pixelfed works, which is to provide an outbox endpoint for the purposes of identifying an actor as an actor (since actors have inbox + outbox) -- but the outbox is not accessible and will return an error if you try to fetch it. bear in mind that the use of outbox to return public activities is only a convention. its primary defined purpose in activitypub is for posting C2S activities, although it "should" return all activities you're authorized to see -- but that's only a "should".

in short, attempting to paginate outboxes is going to be a messy, hackish thing.

**Comment by trwnh** [comment]
@valerauko Are you suggesting paginating the entire outbox from the start?

**Comment by valerauko** [comment]
Exactly! The point here is to make the user's profile "up to date". The outbox is (not sure if by convention or by definition) sorted date descending (so the newest posts come first).

So even just fetching the very first page could make the profile "up to date"

**Comment by trwnh** [comment]
true, outbox would be #34

**Comment by MichaelBell** [comment]
It appears some clients have started using this technique, calling the public /statuses/:id/context endpoint on the remote server.  The problem with this is that if the client renders the toots directly from the context response then it effectively bypasses blocks.  I don't know if any clients are doing that, but it is a concern.

Maybe it would be helpful to have an API call similar to /statuses/:id/context that fetched only enough information to allow the toots to be searched for via the user's local server?  This could then be used explicitly for this purpose, reducing the size of the response from the foreign server, and also reducing the likelihood of admins blocking API access due to concerns about clients bypassing blocks.

**Comment by MichaelBell** [comment]
> there appears to be a sizeable contingent of people who have come to understand blocking to mean denying read access _globally_ rather than merely refusing interaction with their instance

This is one concern, and was part of what led me to suggest a way for a client to read less data from a foreign instance.  But I think the bigger concern here is for user safety.

If a client is fetching toots directly from a foreign instance, then the normal filtering that is applied by the local server is not done.  It's now on the client not to show the user harmful content that the local instance would normally have filtered out for them - either replies from users on an instance that is blocked by the local instance but not the foreign instance, or from users that the fetching user has explicitly blocked.

Therefore, the client should always check to see whether toots that it has pulled from a foreign server are accessible from the local server - but given the public API that allows the client to find the context toots also fetches all the content it would be tempting as a client not to do this.

Clearly a better fix would be to make Mastodon server propagate replies so this functionality was unnecessary, but that is a much more complex fix so client workarounds like this seem inevitable in the short term.

**Comment by audiodude** [comment]
I've created a design doc and mockup for how this might work as a UI only hint:

https://docs.google.com/document/d/1YpdrZVTrk-yt0GpnSxuDybYvFQOA0B-IfkELBYNwSOg/edit?usp=sharing

TLDR:
![image](https://github.com/mastodon/mastodon/assets/57832/31c74389-2e5e-4384-86a1-7e05bf1e688b)

**Comment by audiodude** [comment]
The following issues have been identified as reporting or referencing this problem:

[#14017](https://github.com/mastodon/mastodon/issues/14017): can't see toots on people's profiles or replies to many toots from self-hosted instance
[#18150](https://github.com/mastodon/mastodon/issues/18150): UI button to "see all conversation" using context API
[#9409](https://github.com/mastodon/mastodon/issues/9409): Fetch whole conversation threads
[#34](https://github.com/mastodon/mastodon/issues/34): Backfill statuses from remote accounts when first subscribed
[#17213](https://github.com/mastodon/mastodon/issues/17213): View older posts from other servers
[#3794](https://github.com/mastodon/mastodon/issues/3794): [enhancement] Web UI: Some toots are not shown
